naming/program/src/state.rs

```diff
@@ -0,0 +1,105 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use solana_program::{
+    account_info::AccountInfo,
+    msg,
+    program_error::ProgramError,
+    program_pack::{IsInitialized, Pack, Sealed},
+    pubkey::Pubkey,
+};
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone, Copy)]
+pub enum NameStatus {
+    Uninitialized,
+    Initialized,
+}
+
+/// The data for a Name Registry account is always prefixed a `NameRecordHeader` structure.
+///
+/// The layout of the remaining bytes in the account data are determined by the record `class`
+#[derive(Clone, Debug, BorshSerialize, BorshDeserialize, PartialEq)]
+pub struct NameRecordHeader {
+    // Tells if the name account is initialized.
+    pub name_status: NameStatus,
+
+    // Names are hierarchical.  `parent_name` contains the account address of the parent
+    // name, or `Pubkey::default()` if no parent exists.
+    pub parent_name: Pubkey,
+
+    // The owner of this name
+    pub owner: Pubkey,
+
+    // The class of data this account represents (DNS record, twitter handle, SPL Token name/symbol, etc)
+    //
+    // If `Pubkey::default()` the data is unspecified.
+    pub class: Pubkey,
+}
+
+impl Sealed for NameRecordHeader {}
+
+impl Pack for NameRecordHeader {
+    const LEN: usize = 97;
+
+    fn pack_into_slice(&self, dst: &mut [u8]) {
+        let mut slice = dst;
+        self.serialize(&mut slice).unwrap()
+    }
+
+    fn unpack_from_slice(src: &[u8]) -> Result<Self, ProgramError> {
+        let mut p = src;
+        NameRecordHeader::deserialize(&mut p).map_err(|_| {
+            msg!("Failed to deserialize name record");
+            ProgramError::InvalidAccountData
+        })
+    }
+}
+
+impl IsInitialized for NameRecordHeader {
+    fn is_initialized(&self) -> bool {
+        self.name_status == NameStatus::Initialized
+    }
+}
+
+pub fn write_data(account: &AccountInfo, input: &[u8], offset: u32) {
+    let mut account_data = account.data.borrow_mut();
+    for i in 0..input.len() {
+        account_data[NameRecordHeader::LEN + offset as usize + i] = input[i];
+    }
```

Comment:
It may be more performant to use https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice here instead of re-implementing memcpy